File: src/bun.js/api/crypto/EVP.zig
const EVP = @This();

ctx: BoringSSL.EVP_MD_CTX = undefined,
md: *const BoringSSL.EVP_MD = undefined,
algorithm: Algorithm,

// we do this to avoid asking BoringSSL what the digest name is
// because that API is confusing
pub const Algorithm = enum {
    // @"DSA-SHA",
    // @"DSA-SHA1",
    // @"MD5-SHA1",
    // @"RSA-MD5",
    // @"RSA-RIPEMD160",
    // @"RSA-SHA1",
    // @"RSA-SHA1-2",
    // @"RSA-SHA224",
    // @"RSA-SHA256",
    // @"RSA-SHA384",
    // @"RSA-SHA512",
    // @"ecdsa-with-SHA1",
    blake2b256,
    blake2b512,
    blake2s256,
    md4,
    md5,
    ripemd160,
    sha1,
    sha224,
    sha256,
    sha384,
    sha512,
    @"sha512-224",
    @"sha512-256",

    @"sha3-224",
    @"sha3-256",
    @"sha3-384",
    @"sha3-512",
    shake128,
    shake256,

    pub fn md(this: Algorithm) ?*const BoringSSL.EVP_MD {
        return switch (this) {
            .blake2b256 => BoringSSL.EVP_blake2b256(),
            .blake2b512 => BoringSSL.EVP_blake2b512(),
            .md4 => BoringSSL.EVP_md4(),
            .md5 => BoringSSL.EVP_md5(),
            .ripemd160 => BoringSSL.EVP_ripemd160(),
            .sha1 => BoringSSL.EVP_sha1(),
            .sha224 => BoringSSL.EVP_sha224(),
            .sha256 => BoringSSL.EVP_sha256(),
            .sha384 => BoringSSL.EVP_sha384(),
            .sha512 => BoringSSL.EVP_sha512(),
            .@"sha512-224" => BoringSSL.EVP_sha512_224(),
            .@"sha512-256" => BoringSSL.EVP_sha512_256(),
            else => null,
        };
    }

    pub const names: std.EnumArray(Algorithm, bun.String) = brk: {
        var all = std.EnumArray(Algorithm, bun.String).initUndefined();
        var iter = all.iterator();
        while (iter.next()) |entry| {
            entry.value.* = bun.String.init(@tagName(entry.key));
        }
        break :brk all;
    };

    pub const map = bun.ComptimeStringMap(Algorithm, .{
        .{ "blake2b256", .blake2b256 },
        .{ "blake2b512", .blake2b512 },
        .{ "blake2s256", .blake2s256 },
        .{ "ripemd160", .ripemd160 },
        .{ "rmd160", .ripemd160 },
        .{ "md4", .md4 },
        .{ "md5", .md5 },
        .{ "sha1", .sha1 },
        .{ "sha128", .sha1 },
        .{ "sha224", .sha224 },
        .{ "sha256", .sha256 },
        .{ "sha384", .sha384 },
        .{ "sha512", .sha512 },
        .{ "sha-1", .sha1 },
        .{ "sha-224", .sha224 },
        .{ "sha-256", .sha256 },
        .{ "sha-384", .sha384 },
        .{ "sha-512", .sha512 },
        .{ "sha-512/224", .@"sha512-224" },
        .{ "sha-512_224", .@"sha512-224" },
        .{ "sha-512224", .@"sha512-224" },
        .{ "sha512-224", .@"sha512-224" },
        .{ "sha-512/256", .@"sha512-256" },
        .{ "sha-512_256", .@"sha512-256" },
        .{ "sha-512256", .@"sha512-256" },
        .{ "sha512-256", .@"sha512-256" },
        .{ "sha384", .sha384 },
        .{ "sha3-224", .@"sha3-224" },
        .{ "sha3-256", .@"sha3-256" },
        .{ "sha3-384", .@"sha3-384" },
        .{ "sha3-512", .@"sha3-512" },
        .{ "shake128", .shake128 },
        .{ "shake256", .shake256 },
        // .{ "md5-sha1", .@"MD5-SHA1" },
        // .{ "dsa-sha", .@"DSA-SHA" },
        // .{ "dsa-sha1", .@"DSA-SHA1" },
        // .{ "ecdsa-with-sha1", .@"ecdsa-with-SHA1" },
        // .{ "rsa-md5", .@"RSA-MD5" },
        // .{ "rsa-sha1", .@"RSA-SHA1" },
        // .{ "rsa-sha1-2", .@"RSA-SHA1-2" },
        // .{ "rsa-sha224", .@"RSA-SHA224" },
        // .{ "rsa-sha256", .@"RSA-SHA256" },
        // .{ "rsa-sha384", .@"RSA-SHA384" },
        // .{ "rsa-sha512", .@"RSA-SHA512" },
        // .{ "rsa-ripemd160", .@"RSA-RIPEMD160" },
    });
};

pub fn init(algorithm: Algorithm, md: *const BoringSSL.EVP_MD, engine: *BoringSSL.ENGINE) EVP {
    bun.BoringSSL.load();

    var ctx: BoringSSL.EVP_MD_CTX = undefined;
    BoringSSL.EVP_MD_CTX_init(&ctx);
    _ = BoringSSL.EVP_DigestInit_ex(&ctx, md, engine);
    return .{
        .ctx = ctx,
        .md = md,
        .algorithm = algorithm,
    };
}

pub fn reset(this: *EVP, engine: *BoringSSL.ENGINE) void {
    BoringSSL.ERR_clear_error();
    _ = BoringSSL.EVP_DigestInit_ex(&this.ctx, this.md, engine);
}

pub fn hash(this: *EVP, engine: *BoringSSL.ENGINE, input: []const u8, output: []u8) ?u32 {
    BoringSSL.ERR_clear_error();
    var outsize: c_uint = @min(@as(u16, @truncate(output.len)), this.size());
    if (BoringSSL.EVP_Digest(input.ptr, input.len, output.ptr, &outsize, this.md, engine) != 1) {
        return null;
    }

    return outsize;
}

pub fn final(this: *EVP, engine: *BoringSSL.ENGINE, output: []u8) []u8 {
    BoringSSL.ERR_clear_error();
    var outsize: u32 = @min(@as(u16, @truncate(output.len)), this.size());
    if (BoringSSL.EVP_DigestFinal_ex(
        &this.ctx,
        output.ptr,
        &outsize,
    ) != 1) {
        return "";
    }

    this.reset(engine);

    return output[0..outsize];
}

pub fn update(this: *EVP, input: []const u8) void {
    BoringSSL.ERR_clear_error();
    _ = BoringSSL.EVP_DigestUpdate(&this.ctx, input.ptr, input.len);
}

pub fn size(this: *const EVP) u16 {
    return @as(u16, @truncate(BoringSSL.EVP_MD_CTX_size(&this.ctx)));
}

pub fn copy(this: *const EVP, engine: *BoringSSL.ENGINE) error{OutOfMemory}!EVP {
    BoringSSL.ERR_clear_error();
    var new = init(this.algorithm, this.md, engine);
    if (BoringSSL.EVP_MD_CTX_copy_ex(&new.ctx, &this.ctx) == 0) {
        return error.OutOfMemory;
    }
    return new;
}

pub fn byNameAndEngine(engine: *BoringSSL.ENGINE, name: []const u8) ?EVP {
    if (Algorithm.map.getWithEql(name, strings.eqlCaseInsensitiveASCIIIgnoreLength)) |algorithm| {
        if (algorithm.md()) |md| {
            return EVP.init(algorithm, md, engine);
        }

        if (BoringSSL.EVP_get_digestbyname(@tagName(algorithm))) |md| {
            return EVP.init(algorithm, md, engine);
        }
    }

    return null;
}

pub fn byName(name: ZigString, global: *jsc.JSGlobalObject) ?EVP {
    var name_str = name.toSlice(global.allocator());
    defer name_str.deinit();
    return byNameAndEngine(global.bunVM().rareData().boringEngine(), name_str.slice());
}

pub fn deinit(this: *EVP) void {
    // https://github.com/oven-sh/bun/issues/3250
    _ = BoringSSL.EVP_MD_CTX_cleanup(&this.ctx);
}

pub const Digest = [BoringSSL.EVP_MAX_MD_SIZE]u8;
pub const PBKDF2 = @import("./PBKDF2.zig");
pub const pbkdf2 = PBKDF2.pbkdf2;

const std = @import("std");

const bun = @import("bun");
const strings = bun.strings;
const BoringSSL = bun.BoringSSL.c;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const ZigString = jsc.ZigString;

File: src/runtime/crypto/EVP.rs
use core::ffi::CStr;

use bun_alloc::AllocError;
use bun_boringssl_sys as boringssl;
use bun_core::String as BunString;

use crate::jsc::JSGlobalObject;

pub struct EVP {
    ctx: boringssl::DigestCtx,
    algorithm: Algorithm,
}

// ──────────────────────────────────────────────────────────────────────────
// The `Algorithm` enum + `md()` live in
// `bun_sha_hmac::evp` so lower-tier crates (`bun_csrf`, `bun_sha_hmac::hmac`)
// can name it without depending upward on bun_runtime. Re-export the canonical
// enum here; the higher-tier extras (`names`, `lookup`, `tag_cstr`) that need
// bun_str live below as an extension trait / free fns on the re-exported type.
// ──────────────────────────────────────────────────────────────────────────
pub use bun_sha_hmac::evp::Algorithm;

/// Higher-tier helpers on the lowered `Algorithm` enum (orphan rules prevent an
/// inherent `impl` on a foreign type, so callers `use evp::AlgorithmExt as _;`).
pub(crate) trait AlgorithmExt: Copy + Sized {
    /// NUL-terminated tag name. Needed for `EVP_get_digestbyname` which reads
    /// a C string.
    fn tag_cstr(self) -> &'static CStr;

    /// `bun.String` view of every algorithm tag name; returned as a flat slice
    /// since the enum is foreign and cannot derive `enum_map::Enum`.
    fn names() -> &'static [BunString];
}

impl AlgorithmExt for Algorithm {
    fn tag_cstr(self) -> &'static CStr {
        match self {
            Algorithm::Blake2b256 => c"blake2b256",
            Algorithm::Blake2b512 => c"blake2b512",
            Algorithm::Blake2s256 => c"blake2s256",
            Algorithm::Md4 => c"md4",
            Algorithm::Md5 => c"md5",
            Algorithm::Ripemd160 => c"ripemd160",
            Algorithm::Sha1 => c"sha1",
            Algorithm::Sha224 => c"sha224",
            Algorithm::Sha256 => c"sha256",
            Algorithm::Sha384 => c"sha384",
            Algorithm::Sha512 => c"sha512",
            Algorithm::Sha512_224 => c"sha512-224",
            Algorithm::Sha512_256 => c"sha512-256",
            Algorithm::Sha3_224 => c"sha3-224",
            Algorithm::Sha3_256 => c"sha3-256",
            Algorithm::Sha3_384 => c"sha3-384",
            Algorithm::Sha3_512 => c"sha3-512",
            Algorithm::Shake128 => c"shake128",
            Algorithm::Shake256 => c"shake256",
            // upstream enum is `#[non_exhaustive]`; the variant set is closed in
            // practice.
            _ => unreachable!("unhandled EVP algorithm variant"),
        }
    }

    fn names() -> &'static [BunString] {
        static NAMES: std::sync::OnceLock<[BunString; ALL.len()]> = std::sync::OnceLock::new();
        NAMES
            .get_or_init(|| {
                core::array::from_fn(|i| BunString::static_(ALL[i].tag_cstr().to_bytes()))
            })
            .as_slice()
    }
}

/// Stable iteration order over every `Algorithm` variant — the lowered enum is
/// foreign + `#[non_exhaustive]`, so we can't derive an iterator for it.
const ALL: [Algorithm; 19] = [
    Algorithm::Blake2b256,
    Algorithm::Blake2b512,
    Algorithm::Blake2s256,
    Algorithm::Md4,
    Algorithm::Md5,
    Algorithm::Ripemd160,
    Algorithm::Sha1,
    Algorithm::Sha224,
    Algorithm::Sha256,
    Algorithm::Sha384,
    Algorithm::Sha512,
    Algorithm::Sha512_224,
    Algorithm::Sha512_256,
    Algorithm::Sha3_224,
    Algorithm::Sha3_256,
    Algorithm::Sha3_384,
    Algorithm::Sha3_512,
    Algorithm::Shake128,
    Algorithm::Shake256,
];

/// Algorithm names joined as `"'a', 'b', … 'y' or 'z'"` (declaration order),
/// for "must be one of" error messages.
pub(crate) const ALGORITHM_ONE_OF: &str = "'blake2b256', 'blake2b512', 'blake2s256', 'md4', 'md5', \
'ripemd160', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512', 'sha512-224', 'sha512-256', \
'sha3-224', 'sha3-256', 'sha3-384', 'sha3-512', 'shake128' or 'shake256'";

bun_core::comptime_string_map! {
    /// Case-sensitive name → `Algorithm`. Keys must stay lowercase so
    /// `lookup_ignore_case` can use the ASCII-case-insensitive probe.
    static ALGORITHM_MAP: Algorithm = {
        b"md4" => Algorithm::Md4,
        b"md5" => Algorithm::Md5,
        b"sha1" => Algorithm::Sha1,
        b"sha-1" => Algorithm::Sha1,
        b"sha128" => Algorithm::Sha1,
        b"sha224" => Algorithm::Sha224,
        b"sha256" => Algorithm::Sha256,
        b"sha384" => Algorithm::Sha384,
        b"sha512" => Algorithm::Sha512,
        b"rmd160" => Algorithm::Ripemd160,
        b"sha-224" => Algorithm::Sha224,
        b"sha-256" => Algorithm::Sha256,
        b"sha-384" => Algorithm::Sha384,
        b"sha-512" => Algorithm::Sha512,
        b"sha3-224" => Algorithm::Sha3_224,
        b"sha3-256" => Algorithm::Sha3_256,
        b"sha3-384" => Algorithm::Sha3_384,
        b"sha3-512" => Algorithm::Sha3_512,
        b"shake128" => Algorithm::Shake128,
        b"shake256" => Algorithm::Shake256,
        b"shake-128" => Algorithm::Shake128,
        b"shake-256" => Algorithm::Shake256,
        b"ripemd160" => Algorithm::Ripemd160,
        b"blake2b256" => Algorithm::Blake2b256,
        b"blake2b512" => Algorithm::Blake2b512,
        b"blake2s256" => Algorithm::Blake2s256,
        b"sha-512224" => Algorithm::Sha512_224,
        b"sha512-224" => Algorithm::Sha512_224,
        b"sha-512256" => Algorithm::Sha512_256,
        b"sha512-256" => Algorithm::Sha512_256,
        b"sha-512/224" => Algorithm::Sha512_224,
        b"sha-512_224" => Algorithm::Sha512_224,
        b"sha-512/256" => Algorithm::Sha512_256,
        b"sha-512_256" => Algorithm::Sha512_256,
    };
}
// Aliases the Zig table listed but never wired up:
// b"md5-sha1" => .@"MD5-SHA1",
// b"dsa-sha" => .@"DSA-SHA",
// b"dsa-sha1" => .@"DSA-SHA1",
// b"ecdsa-with-sha1" => .@"ecdsa-with-SHA1",
// b"rsa-md5" => .@"RSA-MD5",
// b"rsa-sha1" => .@"RSA-SHA1",
// b"rsa-sha1-2" => .@"RSA-SHA1-2",
// b"rsa-sha224" => .@"RSA-SHA224",
// b"rsa-sha256" => .@"RSA-SHA256",
// b"rsa-sha384" => .@"RSA-SHA384",
// b"rsa-sha512" => .@"RSA-SHA512",
// b"rsa-ripemd160" => .@"RSA-RIPEMD160",

/// ASCII-case-insensitive name → `Algorithm`.
pub(crate) fn lookup_ignore_case(bytes: &[u8]) -> Option<Algorithm> {
    ALGORITHM_MAP.get_ascii_case_insensitive(bytes).copied()
}

impl EVP {
    pub fn algorithm(&self) -> Algorithm {
        self.algorithm
    }

    pub(crate) fn init(
        algorithm: Algorithm,
        md: &'static boringssl::EVP_MD,
        engine: Option<&boringssl::ENGINE>,
    ) -> EVP {
        bun_boringssl::load();

        EVP {
            ctx: boringssl::DigestCtx::new(md, engine),
            algorithm,
        }
    }

    pub fn reset(&mut self, engine: Option<&boringssl::ENGINE>) {
        boringssl::ERR_clear_error();
        let _ = self.ctx.init(self.ctx.md(), engine);
    }

    pub(crate) fn hash(
        &mut self,
        engine: Option<&boringssl::ENGINE>,
        input: &[u8],
        output: &mut [u8],
    ) -> Option<u32> {
        boringssl::ERR_clear_error();
        boringssl::digest(self.ctx.md(), input, output, engine)
    }

    pub(crate) fn r#final<'a>(
        &mut self,
        engine: Option<&boringssl::ENGINE>,
        output: &'a mut [u8],
    ) -> &'a mut [u8] {
        boringssl::ERR_clear_error();
        let Some(outsize) = self.ctx.final_(output) else {
            return &mut output[..0];
        };

        self.reset(engine);

        &mut output[..outsize as usize]
    }

    pub(crate) fn update(&mut self, input: &[u8]) {
        boringssl::ERR_clear_error();
        let _ = self.ctx.update(input);
    }

    pub(crate) fn size(&self) -> u16 {
        self.ctx.size() as u16
    }

    pub(crate) fn copy(&self, engine: Option<&boringssl::ENGINE>) -> Result<EVP, AllocError> {
        boringssl::ERR_clear_error();
        let mut new = EVP::init(self.algorithm, self.ctx.md(), engine);
        if !new.ctx.copy_from(&self.ctx) {
            return Err(AllocError);
        }
        Ok(new)
    }

    pub(crate) fn by_name_and_engine(
        engine: Option<&boringssl::ENGINE>,
        name: &[u8],
    ) -> Option<EVP> {
        if let Some(algorithm) = lookup_ignore_case(name) {
            if let Some(md) = algorithm.md() {
                return Some(EVP::init(algorithm, md, engine));
            }

            // strum's `<&'static str>::from(algorithm)` is NOT NUL-terminated, so use the
            // explicit `tag_cstr()` table for the C-string lookup.
            if let Some(md) = boringssl::digest_by_name(algorithm.tag_cstr()) {
                return Some(EVP::init(algorithm, md, engine));
            }
        }

        None
    }

    pub(crate) fn by_name(name: &[u8], global: &JSGlobalObject) -> Option<EVP> {
        let engine = global.bun_vm().as_mut().rare_data().boring_engine();
        Self::by_name_and_engine(engine, name)
    }
}

pub(crate) type Digest = [u8; boringssl::EVP_MAX_MD_SIZE as usize];
